### PR TITLE
Linear solvers

### PR DESCRIPTION
**Pull request by mwarusz**
This pull request
- add a LinearSolvers module
- implements one concrete iterative linear solver - Generalized Conjugate Residual
- adds tests for the above

### CODE REVIEW

**Comment by mwarusz** [comment]
bors try

**Comment by mwarusz** [comment]
bors try

**Comment by mwarusz** [comment]
bors try

File: test/LinearSolvers/smallsystem.jl [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+# https://github.com/JuliaMath/IterativeSolvers.jl/blob/master/test/cg.jl
+@testset "LinearSolvers small full system" begin
+  n = 10
+
+  expected_iters = Dict(Float32 => 4, Float64 => 6)
+
+  for T in [Float32, Float64]
+    Random.seed!(44)
+
+    A = rand(T, n, n)
```
Comment (charleskawczynski) [inline test/LinearSolvers/smallsystem.jl]: Should we avoid `rand` in the tests?

Comment (lcw) [reply test/LinearSolvers/smallsystem.jl]: I think it is okay to use `rand` as long as we are setting the seed.

**Pull request merged**